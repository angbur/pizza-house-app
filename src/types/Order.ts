import type { Pizza } from 'types/Pizza';

export type OrderItem = {
  _id: string;
  item: Omit<Pizza, 'categories' | '_id'>;
  quantity: number;
};
