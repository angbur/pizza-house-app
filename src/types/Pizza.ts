export const ingredients = [
  'tomato sauce',
  'pineapple',
  'white sauce',
  'cheese',
  'mozzarella',
  'tomatoes',
  'mushrooms',
  'chili',
  'olives',
  'pepper',
  'corn',
  'onion',
  'salami',
  'bacon',
  'shrimps',
  'red fish',
  'basil',
  'arugula',
  'oregano',
  'parsley',
] as const;
export const categories = ['spicy', 'vegetarian', 'meaty', 'seafood'] as const;

export type Ingredients = typeof ingredients[number];
export type Categories = typeof categories[number];

export type Pizza = {
  _id: string;
  name: string;
  ingredients: Ingredients[];
  categories: Categories[];
  price: number;
};
