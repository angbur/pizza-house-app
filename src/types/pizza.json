{
    "name": "Tipsy Tomato",
    "ingredients":["tomato sauce", "tomatoes", "olives", "mozzarella", "cheese", "parsley", "basil"],
    "price": 20,
    "categories": ["vegetarian"]
},
{
    "name": "Pinza Mare",
    "ingredients": ["shrimps", "red fish", "olives", "white sauce", "oregano", "basil"],
    "price": 26,
    "categories": ["seafood"]
},
{
    "name": "The Crispy Crust",
    "ingredients": ["tomato sauce", "salami", "oregano", "olives", "cheese", "basil"],
    "price": 30,
    "categories": ["meaty"]
},
{
    "name": "Feeling Saucy",
    "ingredients": ["tomato sauce", "pineapple", "corn", "olives", "onion", "arugula"],
    "price": 28,
    "categories": ["vegetarian"]
},
{
    "name": "Mellow",
    "ingredients": ["mushrooms", "onion", "basil", "parsley", "tomato sauce"],
    "price": 24,
    "categories": ["vegetarian"]
},
{
    "name": "Heirloom",
    "ingredients": ["mushrooms", "onion", "basil", "parsley", "white sauce", "olives"],
    "price": 24,
    "categories": ["vegetarian"]
},
{
    "name": "Lighthouse",
    "ingredients": ["white sauce", "cheese", "pepper", "chili", "parsley", "basil"],
    "price": 24,
    "categories": ["vegetarian", "spicy"]
},
{
    "name": "Lighthouse",
    "ingredients": ["tomato sauce", "bacon", "olives", "cheese", "parsley", "basil"],
    "price": 24,
    "categories": ["meaty"]
}