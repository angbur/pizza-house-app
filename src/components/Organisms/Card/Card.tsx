import PriceTag from 'components/Atoms/PriceTag/PriceTag';
import Typography from 'components/Atoms/Typography/Typography';
import ButtonWithInput from 'components/Molecules/ButtonWithInput/ButtonWithInput';
import { addToOrder } from 'components/Pages/Order/orderSlice';
import { ThemeContext } from 'components/Theme/ThemeContext';
import { useAppDispatch } from 'hooks';
import { useContext, useState } from 'react';
import styled from 'styled-components';
import { Pizza } from 'types/Pizza';

type CardProps = {
  pizza: Pizza;
};

const PizzaCard = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: baseline;
  align-items: center;
  text-align: center;
  padding: 1rem 1.4rem;
  border-radius: 8px;
  height: 540px;
  width: 360px;
  gap: 0.8rem;
  border: 2px solid ${(props) => props.theme.palette.secondary};
  &:hover {
    border-color: ${(props) => props.theme.palette.light};
  }
  & > img {
    margin: 1rem;
  }
`;

const CardContent = styled.div`
  height: 80px;
  padding: 0 0.6rem;
  & > h4 span {
    text-transform: capitalize;
  }
`;

const CardActions = styled.div`
  display: flex;
  width: 100%;
  margin-top: 0.3rem;
  justify-content: space-between;
  align-items: end;
`;

const Card = ({ pizza }: CardProps) => {
  const theme = useContext(ThemeContext);
  const [quantity, setQuantity] = useState<number>(1);
  const dispatch = useAppDispatch();

  const addQuantity = () => quantity > 0 && quantity < 20 ? setQuantity(prev=>prev+1) : null;
  const substractQuantity = () => quantity > 1 && quantity < 21 ? setQuantity(prev=>prev-1) : null;
  const handleAddToOrder = () => {
    const {categories, _id, ...rest} = pizza;
    dispatch(addToOrder({item: rest, quantity: quantity}));
  };

  return (
    <PizzaCard theme={theme}>
      <img
        src={require(`assets/image/pizza-${pizza._id}.svg`)}
        height={'192px'}
        alt={`Picture of ${pizza.name} pizza`}
      />
      <PriceTag price={pizza.price} />
      <Typography
        variant='card-title'
        color={'light'}
        style={{ ...{ textTransform: 'uppercase' } }}
      >
        {pizza.name}
      </Typography>
      <CardContent>
        <Typography variant='card-subtitle' color={'light'}>
          {pizza.ingredients.map((i, id) =>
            id != pizza.ingredients.length - 1 ? (
              <span key={`ingredients-${i}`}> {i}, </span>
            ) : (
              <span key={`ingredients-${i}`}> {i}</span>
            ),
          )}
        </Typography>
      </CardContent>
      <CardActions>
        <ButtonWithInput value={quantity} onIncrement={addQuantity} onDecrement={substractQuantity} onClick={handleAddToOrder}/>
        <PriceTag price={pizza.price*quantity} size='sm' />
      </CardActions>
    </PizzaCard>
  );
};

export default Card;
