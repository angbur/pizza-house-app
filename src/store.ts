import { configureStore } from '@reduxjs/toolkit';
import orderReducer from 'components/Pages/Order/orderSlice';
import { pizzaApi } from 'services/pizza';

export const store = configureStore({
  reducer: {
    order: orderReducer,
    [pizzaApi.reducerPath]: pizzaApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(pizzaApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
