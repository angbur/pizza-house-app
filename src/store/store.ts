import { configureStore, Store } from '@reduxjs/toolkit';
import orderReducer from 'components/Pages/Order/orderSlice';
import dialogReducer from 'components/Organisms/Dialog/dialogSlice';
import { pizzaApi } from 'services/pizza';
import orderMiddleware from './middleware/orderMiddleware';

export const store: Store = configureStore({
  reducer: {
    order: orderReducer,
    dialog: dialogReducer,
    [pizzaApi.reducerPath]: pizzaApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(pizzaApi.middleware, orderMiddleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
