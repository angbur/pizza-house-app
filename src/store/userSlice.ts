import { createSlice } from '@reduxjs/toolkit';
import { RootState } from './store';
import { toast } from 'react-toastify';
import { userApi } from 'services/user';

type User = {
  loggedIn: boolean;
  status: string | undefined;
  error: string | undefined;
  userId: string | undefined;
  login: string | undefined;
  token: string | undefined;
};

const initialState: User = {
  loggedIn: false,
  status: 'idle',
  error: undefined,
  userId: undefined,
  login: undefined,
  token: undefined,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logout: (state) => {
      state.loggedIn = false;
      state.userId = undefined;
      state.login = undefined;
      state.token = undefined;
      toast.success('Logged out!');
    },
  },
  extraReducers: (builder) => {
    builder.addMatcher(userApi.endpoints.login.matchFulfilled, (state, { payload }) => {
      state.loggedIn = true;
      state.token = payload.token;
      state.userId = payload.userId;
      state.login = payload.login;
    });
  },
});

export const { logout } = userSlice.actions;

export const selectName = (state: RootState) => state.user.user;

export const isLoggedIn = (state: RootState) => state.user.loggedIn;

export const selectResponseStatus = (state: RootState) => state.user.status;

export const selectResponseError = (state: RootState) => state.user.error;

export const selectLoggedInUser = (state: RootState) => state.user.user;

export default userSlice.reducer;
