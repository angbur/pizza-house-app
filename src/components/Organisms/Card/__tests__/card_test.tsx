import renderer from 'react-test-renderer';
import ThemeProvider from 'components/Theme/ThemeContext';
import { Provider } from 'react-redux';
import { store } from 'store/store';
import Card from '../Card';
import { Pizza } from 'types/Pizza';

const pizza: Pizza = {
  _id: '63b2d3b592b7f0f1c7c567b8',
  name: 'Pinza Mare',
  ingredients: ['shrimps', 'red fish', 'olives', 'white sauce', 'oregano', 'basil'],
  price: 26,
  categories: ['seafood'],
};

const Wrapper = () => {
  return (
    <ThemeProvider>
      <Provider store={store}>
        <Card pizza={pizza} />
      </Provider>
    </ThemeProvider>
  );
};

describe('Card', () => {
  test('should renders card correctly', () => {
    const tree = renderer.create(<Wrapper />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
