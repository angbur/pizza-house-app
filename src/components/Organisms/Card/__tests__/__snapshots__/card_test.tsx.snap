// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should renders card correctly 1`] = `
<div
  className="sc-bqWxrE bhcwCd"
>
  <img
    alt="Picture of Pinza Mare pizza"
    className="pizza-img"
    height="192px"
    src={
      Object {
        "ReactComponent": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "__esModule": true,
        "default": "pizza-63b2d3b592b7f0f1c7c567b8.svg",
      }
    }
  />
  <p
    className="price-tag"
    style={
      Object {
        "color": "#EB853E",
        "fontFamily": "Klepon Ijo",
        "fontSize": "3.1500000000000004rem",
        "letterSpacing": "0",
      }
    }
  >
    26$
  </p>
  <h3
    className="card-title"
    style={
      Object {
        "color": "#FCFBFB",
        "fontFamily": "Klepon Ijo",
        "fontSize": "2rem",
        "fontWeight": "400",
        "letterSpacing": "2px",
        "lineHeight": "1.2",
        "textTransform": "uppercase",
      }
    }
  >
    Pinza Mare
  </h3>
  <div
    className="sc-ksBlkl btdDBw"
  >
    <h4
      className="card-subtitle"
      style={
        Object {
          "color": "#FCFBFB",
          "fontFamily": "Source Sans Pro",
          "fontSize": "0.9rem",
          "fontWeight": "400",
          "lineHeight": "1.6",
        }
      }
    >
      <span>
         
        shrimps
        , 
      </span>
      <span>
         
        red fish
        , 
      </span>
      <span>
         
        olives
        , 
      </span>
      <span>
         
        white sauce
        , 
      </span>
      <span>
         
        oregano
        , 
      </span>
      <span>
         
        basil
      </span>
    </h4>
  </div>
  <div
    className="sc-hBxehG fMTVUj"
  >
    <div
      className="sc-ipEyDJ fkvujW quantity-button"
    >
      <button
        className="sc-bcXHqe sc-eDvSVe feNVH cAJNIx secondary-dark btn-sm button"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "0.8500000000000001rem",
            "padding": "0.52rem 1rem",
          }
        }
      >
        <img
          height={17}
          src={
            Object {
              "ReactComponent": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "__esModule": true,
              "default": "basket.svg",
            }
          }
        />
        Add to order
      </button>
      <input
        className="sc-kDvujY dYXGJW"
        onChange={[Function]}
        type="number"
        value={1}
      />
      <div
        className="sc-jrcTuL ggyCiw quantity-button"
      >
        <button
          onClick={[Function]}
        >
          <img
            height={14}
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "arrow-up-active.svg",
              }
            }
          />
        </button>
        <button
          onClick={[Function]}
        >
          <img
            height={14}
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "arrow-down-active.svg",
              }
            }
          />
        </button>
      </div>
    </div>
    <div
      className="sc-eDWCr kebGnF quantity-button-mobile"
    >
      <button
        className="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
            "borderRadius": "0",
            "fontSize": "0.9500000000000001rem",
            "maxHeight": "80px",
            "maxWidth": "80px",
            "padding": "0.5850000000000001rem 1.125rem",
          }
        }
      >
        <img
          height={20}
          src={
            Object {
              "ReactComponent": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "__esModule": true,
              "default": "plus.svg",
            }
          }
        />
      </button>
      <input
        className="sc-csuSiG jFZSmB"
        style={
          Object {
            "border": "none",
            "borderRadius": "0",
            "maxHeight": "80px",
            "maxWidth": "80px",
          }
        }
        type="number"
        value={1}
      />
      <button
        className="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
            "borderRadius": "0",
            "fontSize": "0.9500000000000001rem",
            "maxHeight": "80px",
            "maxWidth": "80px",
            "padding": "0.5850000000000001rem 1.125rem",
          }
        }
      >
        <img
          height={20}
          src={
            Object {
              "ReactComponent": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "__esModule": true,
              "default": "minus.svg",
            }
          }
        />
      </button>
    </div>
    <p
      className="price-tag"
      style={
        Object {
          "color": "#EB853E",
          "fontFamily": "Klepon Ijo",
          "fontSize": "2.25rem",
          "letterSpacing": "0",
        }
      }
    >
      26$
    </p>
  </div>
</div>
`;
